### PR TITLE
Replace UpdateLastActivity with UpdateLastSeen

### LINKED ISSUE #336: Remove LastActivityTime and introduce a throttled LastSeenTime for connections
The last seen time should be updated much less (not request-based).

### PR DESCRIPTION
**Pull request by gingters**
Closes #336

### CODE REVIEW

File: src/Thinktecture.Relay.Server.Persistence.EntityFrameworkCore/StatisticsService.cs [hunk 1]
```diff
@@ -238,4 +254,42 @@ public async Task CleanUpConnectionsAsync(TimeSpan maxAge, CancellationToken can
 			_logger.LogError(23315, ex, "An error occured while deleting old connections");
 		}
 	}
+
+	[LoggerMessage(23316, LogLevel.Debug,
+		"Updating last seen time of connection {ConnectionId} to {LastSeen} in statistics tracking")]
+	partial void LogUpdateLastSeenTime(string connectionId, DateTimeOffset lastSeen);
+
+	/// <inheritdoc/>
+	public async Task UpdateLastSeenTimeAsync(string connectionId,
```
Comment (thomashilzendegen) [inline src/Thinktecture.Relay.Server.Persistence.EntityFrameworkCore/StatisticsService.cs:263]: This method is not part of the interface and seems to be unused.

File: src/Thinktecture.Relay.Server/Maintenance/StatisticsOptions.cs [hunk 2]
```diff
@@ -21,4 +21,9 @@ public class StatisticsOptions
 	/// Indicates whether to clean up stale or closed connections or not.
 	/// </summary>
 	public bool EnableConnectionCleanup { get; set; } = false;
+
+	/// <summary>
+	/// The time interval in which the connection's last seen time will be updated.
+	/// </summary>
+	public TimeSpan LastConnectionSeenUpdateInterval { get; set; } = TimeSpan.FromMinutes(1);
```
Comment (thomashilzendegen) [inline src/Thinktecture.Relay.Server/Maintenance/StatisticsOptions.cs:28]: This naming is quite unclear, maybe it should be named `ConnectionLastSeenUpdateInterval` instead.

File: src/Thinktecture.Relay.Server/Services/ConnectionStatisticsWriter.cs [hunk 3]
```diff
@@ -30,15 +33,22 @@ await scope.ServiceProvider.GetRequiredService<IStatisticsService>()
 	}
 
 	/// <inheritdoc/>
-	public async Task UpdateLastActivityTimeAsync(string connectionId, CancellationToken cancellationToken = default)
+	public Task UpdateLastSeenTimeAsync(string connectionId, CancellationToken cancellationToken = default)
 	{
-		// TODO debounce writing to DB; keep connectionId and UtcNow in dictionary
-		// write to DB every {configurable duration} (default: 1 minute)
-		// additionally add this as a background service just like server statistics writer
+		_buffer[connectionId] = DateTimeOffset.UtcNow;
```
Comment (thomashilzendegen) [inline src/Thinktecture.Relay.Server/Services/ConnectionStatisticsWriter.cs:38]: This is not thread-safe. Better use `ConcurrentDictionary` instead.

File: src/Thinktecture.Relay.Server/Services/ConnectionStatisticsWriter.cs [hunk 4]
```diff
@@ -30,15 +33,22 @@ await scope.ServiceProvider.GetRequiredService<IStatisticsService>()
...
+		return Task.CompletedTask;
+	}
+
+	/// <inheritdoc/>
+	public async Task WriteLastSeenEntriesAsync(CancellationToken stoppingToken)
+	{
+		var buffer = Interlocked.Exchange(ref _buffer,
```
Comment (thomashilzendegen) [inline src/Thinktecture.Relay.Server/Services/ConnectionStatisticsWriter.cs:45]: Better use `ConcurrentDictionary` instead.

File: src/Thinktecture.Relay.Server/Services/ServerStatisticsWriter.cs [hunk 5]
```diff
@@ -60,7 +66,25 @@ protected override async Task ExecuteAsync(CancellationToken stoppingToken)
...
+		await _originWriter.SetShutdownTimeAsync(_serverContext.OriginId, cancellationToken);
 		await base.StopAsync(cancellationToken);
 	}
+
+	private async Task ExecuteOriginUpdatesAsync(CancellationToken stoppingToken)
+	{
+		while (!stoppingToken.IsCancellationRequested)
+		{
+			await _originWriter.UpdateLastSeenTimeAsync(_serverContext.OriginId, stoppingToken);
+			await Task.Delay(_statisticsOptions.LastActivityUpdateInterval, stoppingToken);
```
Comment (thomashilzendegen) [inline src/Thinktecture.Relay.Server/Services/ServerStatisticsWriter.cs:78]: The naming of `LastActivityUpdateInterval` should be changed to `OriginLastSeenUpdateInterval`.

**Review by thomashilzendegen — CHANGES_REQUESTED** [review]

**Comment by gingters** [comment]
Rebased and changed.

File: src/Thinktecture.Relay.Server/Services/ConnectionStatisticsWriter.cs (hunk 3 above)
Comment (thomashilzendegen) [reply src/Thinktecture.Relay.Server/Services/ConnectionStatisticsWriter.cs:38]: This seems not really resolved?

Comment (gingters) [reply src/Thinktecture.Relay.Server/Services/ConnectionStatisticsWriter.cs:38]: This is using the `ConcurrentDictionary` now.

File: src/Thinktecture.Relay.Server/Services/ServerStatisticsWriter.cs (hunk 5 above)
Comment (thomashilzendegen) [reply src/Thinktecture.Relay.Server/Services/ServerStatisticsWriter.cs:78]: It is still named `LastActivityUpdateInterval`?

File: src/Thinktecture.Relay.Server/Services/ConnectionStatisticsWriter.cs (hunk 4 above)
Comment (thomashilzendegen) [reply src/Thinktecture.Relay.Server/Services/ConnectionStatisticsWriter.cs:45]: :/

File: src/Thinktecture.Relay.Server/Maintenance/StatisticsOptions.cs (hunk 2 above)
Comment (thomashilzendegen) [reply src/Thinktecture.Relay.Server/Maintenance/StatisticsOptions.cs:28]: Where is the rename?

**Review by thomashilzendegen — CHANGES_REQUESTED** [review]
Still unresolved things remaining.

**Comment by gingters** [comment]
Nu aber

**Review by thomashilzendegen — APPROVED** [review]

**Pull request merged**